pkg/auth/auth.go

```diff
@@ -0,0 +1,240 @@
+	}
+	return u
+}
+
+// Visas is used to draw the response bytes to a struct
+type Visas struct {
+	Visa []string `json:"ga4gh_passport_v1"`
+}
+
+// VerifyJWT verifies the token signature
+func VerifyJWT(o OIDCDetails, token string) (jwt.Token, string) {
+	log.Debug("verifying JWT signature")
+	// Why do we use context.TODO() ? https://pkg.go.dev/context#TODO
+	keyset, err := jwk.Fetch(context.TODO(), o.JWK)
+	if err != nil {
+		log.Errorf("failed to request JWK set, %s", err)
+		return nil, "aai"
+	}
+	verifiedToken, err := jwt.Parse([]byte(token), jwt.WithKeySet(keyset))
+	if err != nil {
+		log.Errorf("failed to verify token signature, %s", err)
+		return nil, "token"
+	}
+	log.Debug("JWT signature verified")
+	return verifiedToken, ""
+}
+
+// ValidateJWT validates the token claims
+func ValidateJWT(o OIDCDetails, token jwt.Token) bool {
+	log.Debug("validating JWT claims")
+	// claims validated implicitly: iat, exp, nbf
+	// claims validated explicitly: iss
+	// if issuer is set, validate issuer (the case for access token)
+	if len(o.Issuer) > 0 {
+		if err := jwt.Validate(token, jwt.WithIssuer(o.Issuer)); err != nil {
+			log.Errorf("failed to validate token claims, %s", err)
+			return false
+		}
+	} else {
+		// else, don't validate issuer (the case for visas)
+		if err := jwt.Validate(token); err != nil {
+			log.Errorf("failed to validate token claims, %s", err)
+			return false
+		}
+	}
+	log.Debug("JWT claims validated")
+	return true
+}
+
+// GetToken parses the token string from header
+func GetToken(header string) (string, int) {
+	if len(header) == 0 {
+		log.Debug("authorization check failed")
+		return "access token must be provided", 401
+	}
+
+	// Check that Bearer scheme is used
+	headerParts := strings.Split(header, " ")
+	if headerParts[0] != "Bearer" {
+		log.Debug("authorization check failed")
+		return "authorization scheme must be bearer", 400
+	}
+
+	// Check that header contains a token string
+	var token string
+	if len(headerParts) == 2 {
+		token = headerParts[1]
+	} else {
+		log.Debug("authorization check failed")
+		return "token string is missing from authorization header", 400
+	}
+	return token, 0
+}
+
+type JKU struct {
+	URL string `json:"jku"`
+}
+
+// Visa is used to draw the dataset name out of the visa
+type Visa struct {
+	Type    string `json:"type"`
+	Dataset string `json:"value"`
+}
+
+// GetVisas requests the list of visas from userinfo endpoint
+func GetVisas(o OIDCDetails, token string) (bool, []byte) {
+	// Set headers
+	headers := map[string]string{}
+	headers["Authorization"] = "Bearer " + token
+	// Do request
+	code, body, err := request.Do(o.Userinfo, headers)
+	if code != 200 || err != nil {
+		log.Errorf("request failed, %d, %s", code, err)
+		return false, []byte{}
+	}
+	return true, body
+}
+
+// GetPermissions parses visas and finds matching dataset names from the database, returning a list of matches
+func GetPermissions(visas []byte) []string {
```

Comment:
Seems like this can fail and should be able to signal errors as well.

Reply:
Added error handling https://github.com/neicnordic/sda-download/pull/15/commits/0c779dd772fac451b4313e3c7419eb45ccd2285a